### ISSUE TITLE
Wrong router gradient with `experts_implementation="batched_mm"` under expert parallelism

### ISSUE THREAD

**Issue by qgallouedec** [huggingface/transformers#48687]
### System Info

- transformers: main (c694707483)
- torch 2.9, 2x H100, nccl
- Also reproduces on CPU with gloo.

### Who can help?

@ArthurZucker @IlyasMoutawwakil

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

EP with `experts_implementation="batched_mm"` gives the router the wrong gradient. The loss is correct, only the gradient is wrong, so nothing in a training run looks off.

```python
# torchrun --nproc_per_node 2 repro.py
import os, tempfile, torch, torch.distributed as dist
from transformers import OlmoeConfig, OlmoeForCausalLM, set_seed
from transformers.distributed import DistributedConfig

dist.init_process_group("nccl")
rank, world = dist.get_rank(), dist.get_world_size()
torch.cuda.set_device(rank)
device = f"cuda:{rank}"

config = OlmoeConfig(vocab_size=99, hidden_size=32, intermediate_size=16, num_hidden_layers=2,
                     num_attention_heads=4, num_experts=8, num_experts_per_tok=2, norm_topk_prob=False)

path = os.path.join(tempfile.gettempdir(), "ep-router-grad")
if rank == 0:
    set_seed(0)
    OlmoeForCausalLM(config).save_pretrained(path, save_original_format=True)
dist.barrier()

model_ep = OlmoeForCausalLM.from_pretrained(
    path, distributed_config=DistributedConfig(tp_size=world, enable_expert_parallel=True))
model_ref = OlmoeForCausalLM.from_pretrained(path).to(device)
for m in (model_ep, model_ref):
    m.train()
    m.set_experts_implementation("batched_mm")   # grouped_mm is fine

torch.manual_seed(0)
input_ids = torch.randint(0, config.vocab_size, (2, 64), device=device)
labels = torch.randint(0, config.vocab_size, (2, 64), device=device)

model_ref(input_ids, labels=labels).loss.backward()
model_ep(input_ids, labels=labels).loss.backward()

if rank == 0:
    ref = dict(model_ref.named_parameters())
    g = model_ep.get_parameter("model.layers.0.mlp.gate.weight").grad
    g = g.full_tensor() if hasattr(g, "full_tensor") else g
    r = ref["model.layers.0.mlp.gate.weight"].grad
    print(f"router grad norm: reference {r.norm():.6e}  EP {g.norm().to(r.device):.6e}")
```

### Outcome

```
router grad norm: reference 2.051176e-04  EP 3.139034e-04
```

The losses match exactly. Only `mlp.gate.weight` is wrong, on every layer, and it is ~50% too large. Every other parameter agrees to 1e-5.

### Expected behavior

The router gradient should match the non-EP reference, as it does with `grouped_mm`.

### Cause

`batched_mm_experts_forward` clamps the EP sentinel slots into a valid expert and relies on their routing weight being zero:

https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/moe.py#L128-L136

```python
expert_ids = expert_ids.clamp(0, self.num_experts - 1)
...
weighted_out = proj_out * sample_weights.unsqueeze(-1)
```

That is right in the forward: the weight is zero so the slot contributes nothing. It is wrong in the backward!, because the gradient of that product with respect to `sample_weights` is `proj_out`, and `proj_out` at a sentinel slot is not zero, it is expert 0 run on that token. So every slot the rank never routed still sends gradient into `top_k_weights`, and from there into the router.

`grouped_mm_experts_forward` in the same file already handles this, masking `proj_out` at both matmuls before the weighted multiply:

```python
sentinel_mask = (expert_ids_g >= self.num_experts).unsqueeze(-1)
...
proj_out = proj_out.masked_fill(sentinel_mask, 0.0)
```

`batched_mm` is just missing it.

Note the EP tests cannot catch this today: `_test_ep_backward_impl` only compares the loss, which is unaffected. #48518 changes it to compare per-parameter gradients, and that is what surfaced this. PR coming

**Comment by qgallouedec** [comment]
Correction on the numbers: the block under Outcome came from the tiny-model test rather than from the script above it. Running the script as posted gives

```
router grad norm: reference 2.240401e-04  EP 2.846920e-04
```

Same finding, and `eager` is fine too, not only `grouped_mm`: on 2 GPUs the router gradient matches the reference for `eager` and `grouped_mm` and only `batched_mm` differs. The losses are identical in all three.
